### ISSUE TITLE
gnt-instance reinstall doesn't update config.data

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1246]
Originally reported of Google Code with ID 1193.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

root@m:~/ganeti# gnt-cluster version
Software version: 2.18.0~alpha1
Internode protocol: 2180000
Configuration format: 2180000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version v2.17.0beta1-252-g8ff1ce9
root@m:~/ganeti# gnt-cluster --version
gnt-cluster (ganeti v2.17.0beta1-252-g8ff1ce9) 2.18.0~alpha1
root@m:~/ganeti# hspace --version
hspace (ganeti) version v2.17.0beta1-252-g8ff1ce9
compiled with ghc 7.6
running on linux x86_64

<b>What distribution are you using?</b>

root@snf-728370:~/ganeti# lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.6 (jessie)
Release:	8.6
Codename:	jessie

<b>What steps will reproduce the problem?</b>
1. gnt-instance add -o snf-image+default --os-parameters img_passwd=passw0rd,img_format=diskdump,img_id=debian_base-8.0-x86_64 -t plain --disk 0:size=2G --no-name-check --no-ip-check -n dio.test.gnt.grnet.gr tsiou
2. gnt-instance shutdown --timeout=0 tsiou
3. gnt-instance reinstall -O img_passwd=f00 tsiou

<b>What is the expected output? What do you see instead?</b>

Expected output:

root@dio:~# gnt-instance info early_change2
- Instance name: early_change2
  UUID: a6921462-81c4-48e7-ae9d-38ac28aaf937
  Serial number: 7
  Creation time: 2016-11-09 17:16:16
  Modification time: 2016-11-09 17:19:59
  State: configured to be down, actual state is down
  Nodes: 
    - primary: dio.test.gnt.grnet.gr
      group: default (UUID a034f01f-90b8-4b4a-9934-fa515135abd4)
    - secondaries: 
  Operating system: snf-image+default
  Operating system parameters: 
    img_format: diskdump
    img_id: debian_base-8.0-x86_64
    img_passwd: f00
...

However, I see:

root@dio:~# gnt-instance info early_change2
- Instance name: early_change2
  UUID: a6921462-81c4-48e7-ae9d-38ac28aaf937
  Serial number: 7
  Creation time: 2016-11-09 17:16:16
  Modification time: 2016-11-09 17:19:59
  State: configured to be down, actual state is down
  Nodes: 
    - primary: dio.test.gnt.grnet.gr
      group: default (UUID a034f01f-90b8-4b4a-9934-fa515135abd4)
    - secondaries: 
  Operating system: snf-image+default
  Operating system parameters: 
    img_format: diskdump
    img_id: debian_base-8.0-x86_64
    img_passwd: passw0rd
...

The password was changed successfully (checked by logging in through a VNC console) but the OS params reported by gnt-instance info remain unchanged.

<b>Please provide any additional information below.</b>

Even worse:

Trying to reinstall an instance twice might result in unexpected behavior:

1. gnt-instance add -o snf-image+default --os-parameters img_passwd=passw0rd,img_format=diskdump,img_id=debian_base-8.0-x86_64 -t plain --disk 0:size=2G --no-name-check --no-ip-check -n dio.test.gnt.grnet.gr tsiou2
2. gnt-instance reinstall -O img_id=ubuntu_desktop-16.04-x86_64
3. gnt-instance reinstall

The second reinstall installs debian_base-8.0-x86_64 back because that's what the config.data knows about the img_id OS param!

Note: The OS provider doesn't really make any difference; I had snf-image "ready" so that's what I used.

```
Originally added on 2016-11-15 09:39:11 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
s/early_change2/tsiou in the command output section. Sorry for that! :)
```
Originally added on 2016-11-15 09:41:08 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Possibly relates to #913
```
Originally added on 2016-11-16 08:19:55 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
This is an interesting one... 

I'm not entirely sure if this is a bug or a "feature". It depends on what we expect from the state of the world vs the state of the record. Should reinstalling an instance with modified OS parameters also change its state of the record? 

I would agree that this looks counter intuitive and I'd also consider it a bug, but I'm not sure if that's how it was originally perceived.

I've looked at the code from 2.10, 2.16 and 2.18 (latest) and they all seem to be affected by this (did not test it on 2.10 but I did test on 2.16 and 2.18 clusters and verified this issue). I'll be digging a bit more into it and report back.
```
Originally added on 2016-11-22 14:29:42 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Well, reinstalling an instance means keep the hardware intact and allow the OS provider (which may be different from the original one!) to flush the disks out and redo the installation. I cannot think of a reason why the user wouldn't want to have the new state of the OS.
```
Originally added on 2016-11-22 16:04:45 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I've already checked the code in 2.10, 2.12, 2.15, 2.16, latest and it 's like this everywhere. My guess is that noone really used(/uses?) the gnt-instance reinstall command; that explains that there are not so many tests in the codebase...

Anyway, another inconsistency would be that the LUInstanceReinstall.Exec() function updates the metadata while it doesn't update the config.

For us (in GRNET), it's definitely a bug.
```
Originally added on 2016-11-24 09:32:49 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fix landed in a789817c6eba26658d7b31ff5c276a8cc07b999a.
Thanks!
```
Originally added on 2016-11-30 10:31:38 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2016-12-02 15:50:46 +0000 UTC.

Changed State: Fixed
